File: app/src/main/java/com/xephorium/crystalnote/data/NoteRepository.java
package com.xephorium.crystalnote.data;

import android.content.Context;
import android.os.Environment;

import com.xephorium.crystalnote.data.model.Note;
import com.xephorium.crystalnote.data.util.NoteUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/*
  NoteRepository.java                                   04.11.2017
  Christopher Cruzen

    Manages read/write to note files in Android's external storage.
*/

public class NoteRepository {


    /*--- Local Variables ---*/

    private final String DIRECTORY_NAME     = "notes";
    private final String FILE_EXTENSION     = ".txt";
    private final String DEFAULT_NOTE_NAME  = "Note";
    private final String DEFAULT_NOTE_TEXT  = "• List Item #5\n• List Item #6\n\n• List Item #7\n• List Item #8\n\nLorem ipsum dolor sit amet, consecte adipiscing elit, sed do eiusmod tempor.";
    private final Context context;


    /*--- Constructor ---*/

    public NoteRepository(Context context) {
        this.context = context;
        createNotesDirectory();
    }


    /*--- Read/Write Methods ---*/

    public List<Note> getNotes() {
        File[] notesList = getNotesDirectory().listFiles();

        if(notesList.length == 0) {
            writeToNote(DEFAULT_NOTE_NAME, DEFAULT_NOTE_TEXT);
            notesList = getNotesDirectory().listFiles();
        }

        List<Note> notes = new ArrayList<>();
        for (int x = 0; x < notesList.length; x++) {
            Note note = new Note();
            note.setName(getNoteName(notesList[x]));
            note.setDate(getNoteDate(notesList[x]));
            note.setPath(getNotePath(notesList[x]));
            note.setPreview(getNotePreview(notesList[x]));
            note.setColor(NoteUtils.getDefaultColor()); // NoteUtils.getRandomColor()
            notes.add(note);
        }

        return notes;
    }

    public Note getNote(String name) {
        List<Note> notes = getNotes();
        for (int x = 0; x < notes.size(); x++) {
            if (notes.get(x).getName().equals(name))
                return notes.get(x);
        }
        return null;
    }

    public String getNoteContents(File note)  {
        if(!note.exists()) {
            return "";
        }

        try {
            InputStream inputStream = new FileInputStream(note.toString());

            if(inputStream != null) {
                InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
                BufferedReader bufferedReader = new BufferedReader(inputStreamReader);
                String nextLine = "";
                StringBuilder stringBuilder = new StringBuilder();

                while ((nextLine = bufferedReader.readLine()) != null ) {
                    stringBuilder.append(nextLine);
                    stringBuilder.append("\n");
                }

                inputStream.close();
                return stringBuilder.toString();
            }
            return "";
        }
        catch (Exception e) {
            return "";
        }
    }

    public boolean writeToNote(String name, String content) {
        File note = getNoteFile(name);

        if(!note.exists())
            createNote(name);

        try {
            FileOutputStream noteOutputStream = new FileOutputStream(note);
            PrintWriter printWriter = new PrintWriter(noteOutputStream);
            printWriter.print(content);
            printWriter.flush();
            printWriter.close();
            noteOutputStream.close();
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean createNote(String name) {
        File notesFile = new File(getNotesDirectory().toString() + "/" + name + FILE_EXTENSION);

        if(notesFile == null)
            return false;

        if(!notesFile.exists()) {
            try {
                return notesFile.createNewFile();
            } catch (Exception e) {
                return false;
            }
        } else {
            return true;
        }
    }

    public File getNoteFile(Note note) {
        return new File(getNotesDirectory().toString() + "/" + note.getName() + FILE_EXTENSION);
    }

    public boolean noteExists(String name) {
        List<Note> notes = getNotes();
        for (int x = 0; x < notes.size(); x++) {
            if (notes.get(x).getName().equals(name))
                return true;
        }
        return false;
    }

    private File getNoteFile(String name) {
        return new File(getNotesDirectory().toString() + "/" + name + FILE_EXTENSION);
    }

    private String getNoteName(File note) {
        return note.toString().substring(getNotesDirectory().toString().length() + "/".length(), (int) note.toString().length() - FILE_EXTENSION.length());
    }

    private Date getNoteDate(File note) {
        if(!note.exists()) {
            return null;
        }
        return new Date(note.lastModified());
    }

    private String getNotePath(File note) {
        if(!note.exists()) {
            return "";
        }
        return note.getPath();
    }

    private String getNotePreview(File note) {
        if(!note.exists()) {
            return "";
        }
        return NoteUtils.getPreview(getNoteContents(note));
    }


    /*--- Setup & Utility Methods ---*/

    public boolean createNotesDirectory() {
        File directory = getNotesDirectory();

        if(directory == null)
            return false;

        if(!directory.exists()) {
            if(!directory.mkdir())
                return false;
            else
                populateNotesDirectory();
            return true;
        } else {
            if(notesDirectoryIsEmpty())
                populateNotesDirectory();
            return true;
        }
    }

    public boolean notesDirectoryIsEmpty() {
        return (getNotesDirectory().listFiles().length == 0);
    }

    public File getNotesDirectory() {
        File tempDirectory = context.getExternalFilesDir(Environment.DIRECTORY_DCIM);
        if(tempDirectory != null) {
            tempDirectory.delete();
            (new File(tempDirectory.toString().substring(0, (int) (tempDirectory.toString().length() - "/DCIM".length())))).delete();
            return new File(tempDirectory.toString().substring(0, (int) (tempDirectory.toString().length() - "/files/DCIM".length())) + "/" + DIRECTORY_NAME);
        } else
            return null;
    }

    private void populateNotesDirectory() {
        writeToNote(DEFAULT_NOTE_NAME, DEFAULT_NOTE_TEXT);

        // Test Data
        writeToNote("Shopping List", "- Bread\n- Milk\n- Eggs\n- Sugar");
        writeToNote("Beautiful Vistas", "- Shipyard, Halo Reach\n- All of Skyrim\n- Winterfell, The Forest");
        writeToNote("Reasons To Code", "- Puzzle Solving!\n- Free To Learn\n- Hones Critical Thinking\n- It Pays Well");
        writeToNote("Cities to Visit", "- Chicago\n- New York\n- Seattle\n- San Francisco");
        writeToNote("A Final Note", "Building this app has been challenging. There were more than a few moments when I was ready to throw in the towel and accept that I'd learned enough. But each successfully rendered list item, formatted to responsive perfection has been a reminder of why I started: to create something beautiful. The world deserves a clean notes app for Android. It's about time someone put in the hours to code one.");
    }
}

File: app/src/main/java/com/xephorium/crystalnote/data/NoteRepository.kt
package com.xephorium.crystalnote.data

import android.content.Context
import android.os.Environment

import com.xephorium.crystalnote.data.model.Note
import com.xephorium.crystalnote.data.utility.NoteUtility

import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStreamReader
import java.io.PrintWriter
import java.util.ArrayList
import java.util.Date

/*
  NoteRepository                                        05.11.2019
  Christopher Cruzen

    Manages read/write to note files in Android's external storage.
*/

class NoteRepository(private val context: Context) {


    /*--- Object Initialization ---*/

    init {
        createNotesDirectory()
    }


    /*--- Public Read/Write Methods ---*/

    fun getNotes(): List<Note> {
        val notesList = getNotesDirectory().listFiles()

        //if (notesList.isEmpty()) {
        //    writeToNote(DEFAULT_NOTE_NAME, DEFAULT_NOTE_TEXT)
        //    notesList = getNotesDirectory().listFiles()
        //}

        val notes = mutableListOf<Note>()
        notesList.forEach { file ->
            notes.add(Note(
                    color = NoteUtility.getDefaultColor(),
                    name = getNoteName(file),
                    date = getNoteDate(file),
                    path = getNotePath(file),
                    preview = getNotePreview(file)
            ))
        }

        return notes
    }

    fun getNote(name: String): Note? {
        return getNotes().firstOrNull { currentNote -> currentNote.name == name }
    }

    fun writeToNote(name: String, content: String): Boolean {
        val note = getNoteFile(name)

        if (!note.exists())
            createNote(name)

        try {
            val noteOutputStream = FileOutputStream(note)
            val printWriter = PrintWriter(noteOutputStream)
            printWriter.print(content)
            printWriter.flush()
            printWriter.close()
            noteOutputStream.close()
            return true
        } catch (e: Exception) {
            return false
        }
    }

    fun readNoteContents(name: String): String {
        val note = getNoteFile(name)
        if (!note.exists()) return ""

        try {
            val inputStream = FileInputStream(note.toString())
            val inputStreamReader = InputStreamReader(inputStream)
            val bufferedReader = BufferedReader(inputStreamReader)
            val stringBuilder = StringBuilder()

            var nextLine = bufferedReader.readLine()
            while (nextLine != null) {
                stringBuilder.append(nextLine)
                stringBuilder.append("\n")
                nextLine = bufferedReader.readLine()
            }
            inputStream.close()

            return stringBuilder.toString()

        } catch (e: Exception) {
            return ""
        }
    }

    fun deleteNote(name: String) {
        if (noteExists(name)) {
            try {
                getNoteFile(name).delete()
            } catch (e: Exception) {
                // Do Nothing
            }
        }
    }

    fun noteExists(name: String): Boolean {
        getNotes().forEach { note ->
            if (note.name == name) return true
        }
        return false
    }


    /*--- Private File Management Methods ---*/

    private fun createNote(name: String) {
        if (!noteExists(name)) {
            try {
                getNoteFile(name).createNewFile()
            } catch (e: Exception) {
                // Do Nothing
            }
        }
    }

    private fun getNoteFile(note: Note): File {
        return File(getNotesDirectory().toString() + "/" + note.name + FILE_EXTENSION)
    }

    private fun getNoteFile(name: String): File {
        return File(getNotesDirectory().toString() + "/" + name + FILE_EXTENSION)
    }

    private fun getNoteName(note: File): String {
        return note.toString().substring(getNotesDirectory().toString().length
                + "/".length, note.toString().length - FILE_EXTENSION.length)
    }

    private fun getNoteDate(note: File): Date {
        return Date(note.lastModified())
    }

    private fun getNotePath(note: File): String {
        return note.path
    }

    private fun getNotePreview(note: File): String {
        return NoteUtility.getPreview(readNoteContents(getNoteName(note)))
    }


    /*--- Private Directory Management Methods ---*/

    /* Note: Method returns a File object pointing to the external device directory
     *       <app home>/notes. When passed the parameter DIRECTORY_DCIM, getExternalFilesDir()
     *       returns <app home>/files/DCIM. This method then extracts <app home>, deletes
     *       the newly created DCIM and files folders, and returns the final notes directory.
     */
    private fun getNotesDirectory(): File {
        return context.getExternalFilesDir(Environment.DIRECTORY_DCIM)!!.let {

            // Delete DCIM Directory
            it.delete()

            // Delete files Directory
            val filesDirectory =
                    it.toString().substring(0, it.toString().length - "/DCIM".length)
            File(filesDirectory).delete()

            // Assemble & Return Notes Directory
            val notesDirectory =
                    it.toString().substring(0, it.toString().length - "/files/DCIM".length) +
                            "/" + DIRECTORY_NAME
            File(notesDirectory)
        }
    }

    private fun createNotesDirectory() {
        val directory = getNotesDirectory()
        if (!directory.exists()) directory.mkdir()
    }

    private fun isNotesDirectoryEmpty(): Boolean {
        return getNotesDirectory().listFiles().isEmpty()
    }

    private fun populateNotesDirectory() {
        writeToNote(DEFAULT_NOTE_NAME, DEFAULT_NOTE_TEXT)

        // Test Data
        writeToNote("Shopping List", "- Bread\n- Milk\n- Eggs\n- Sugar")
        writeToNote("Beautiful Vistas", "- Shipyard, Halo Reach\n- All of Skyrim\n- Winterfell, The Forest")
        writeToNote("Reasons To Code", "- Puzzle Solving!\n- Free To Learn\n- Hones Critical Thinking\n- It Pays Well")
        writeToNote("Cities to Visit", "- Chicago\n- New York\n- Seattle\n- San Francisco")
        writeToNote("A Final Note", "Building this app has been challenging. There were more than a few moments when I was ready to throw in the towel and accept that I'd learned enough. But each successfully rendered list item, formatted to responsive perfection has been a reminder of why I started: to create something beautiful. The world deserves a clean notes app for Android. It's about time someone put in the hours to code one.")
    }


    /*--- Constants ---*/

    companion object {
        private const val DIRECTORY_NAME = "notes"
        private const val FILE_EXTENSION = ".txt"
        private const val DEFAULT_NOTE_NAME = "Note"
        private const val DEFAULT_NOTE_TEXT = "• List Item #5\n• List Item #6\n\n• List Item #7\n• List Item #8\n\nLorem ipsum dolor sit amet, consecte adipiscing elit, sed do eiusmod tempor."
    }
}
